File: GoldMan Sachs/GroupAnagrams/GroupAnagrams.cpp
// { Driver Code Starts
// Initial Template for C++
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution
{
public:
    vector<vector<string>> Anagrams(vector<string> &words)
    {
        unordered_map<string, vector<string>> hm;

        for (string &s : words)
        {
            string r = s;
            sort(r.begin(), r.end());
            hm[r].push_back(s);
        }

        vector<vector<string>> ans;

        for (const pair<string, vector<string>> &p : hm)
        {
            ans.push_back(p.second);
        }

        return ans;
    }
};

// { Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;
        cin >> n;
        vector<string> string_list(n);
        for (int i = 0; i < n; ++i)
        {
            cin >> string_list[i];
        }
        Solution ob;
        vector<vector<string>> result = ob.Anagrams(string_list);
        sort(result.begin(), result.end());
        for (int i = 0; i < result.size(); i++)
        {
            for (int j = 0; j < result[i].size(); j++)
            {
                cout << result[i][j] << " ";
            }
            cout << "\n";
        }
    }

    return 0;
}
// } Driver Code Ends

File: GoldMan Sachs/GroupAnagrams/GroupAnagrams.java
// { Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class GroupAnagrams {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine().trim());
        while (t > 0) {
            int n = Integer.parseInt(br.readLine().trim());
            String x = br.readLine().trim();
            String string_list[] = x.split(" ", n);

            Solution ob = new Solution();

            List<List<String>> ans = ob.Anagrams(string_list);

            Collections.sort(ans, new Comparator<List<String>>() {
                public int compare(List<String> l1, List<String> l2) {
                    String s1 = l1.get(0);
                    String s2 = l2.get(0);

                    return s1.compareTo(s2);
                }
            });

            for (int i = 0; i < ans.size(); i++) {
                for (int j = 0; j < ans.get(i).size(); j++) {
                    System.out.print(ans.get(i).get(j) + " ");
                }
                System.out.println();
            }

            t--;
        }
    }

}
// } Driver Code Ends

// User function Template for Java

class Solution {
    public List<List<String>> Anagrams(String[] words) {
        HashMap<String, List<String>> hm = new HashMap<>();

        for (String s : words) {
            char[] arr = s.toCharArray();
            Arrays.sort(arr);

            String p = Arrays.toString(arr);

            List<String> strs = hm.getOrDefault(p, new ArrayList<>());
            strs.add(s);
            hm.put(p, strs);
        }

        List<List<String>> ans = new ArrayList<>();

        for (String key : hm.keySet()) {
            ans.add(hm.get(key));
        }

        return ans;
    }
}
